scipy/stats/_axis_nan_policy.py

```diff
@@ -226,6 +226,26 @@ def _check_empty_inputs(samples, axis):
     return output
 
 
+def _add_reduced_axes(res, reduced_axes, n_outputs, keepdims,
+                      result_object):
+    """
+    Add reduced axes back to all the arrays in the result object
+    if keepdims = True. `res` should either be an instance of the
+    result_object or a tuple of results.
+    """
+    if not keepdims:
+        if n_outputs == 1:
+            return result_object(res)
+        return result_object(*res)
+    res_expanded = []
+    if n_outputs == 1:
+        res_expanded.append(np.expand_dims(res, reduced_axes))
+    else:
+        for r in res:
+            res_expanded.append(np.expand_dims(r, reduced_axes))
+    return result_object(*res_expanded)
```

Comment:
We need to work on the rest of this until this function is just:
```suggestion
    res = ([np.expand_dims(output, reduced_axes) for output in res] if keepdims
           else res)
    return result_object(*res)
```
And we don't want to have to do things like 
```
            if n_outputs > 1:
                res = result_unpacker(res)
```
In some places but not others. If we do it in one place, we should do it everywhere. I think we can get this so we always apply the `result_unpacker`.

With some slight tweaks to other parts, I'm sure we can get there. 

For the ndarray part of the decorator, the reason it's tricky is that `result_unpacker` wasn't designed to work with the raw output of a function when we use the function's native vectorization. It was only designed to work with output of a function that _we_ vectorized using `np.apply_along_axis`. Roughly speaking, for multi-output functions: 
- when we use the native vectorization, the different outputs come along the zeroth axis (because the output is really a tuple of arrays)
- when we vectorize using `np.apply_along_axis`, the different outputs come along the _last_ axis because I standardized the statistic to be taken along the last axis. The fix is probably to standardize to take the statistic along the zeroth axis. I ran into this same sort of thing recently in gh-15491.

Give me a few more days to play with this.

Reply:
> * When we vectorize using `np.apply_along_axis`, the different outputs come along the _last_ axis because I standardized the statistic to be taken along the last axis. The fix is probably to standardize to take the statistic along the zeroth axis.

That sounds good. I was hesitant to touch that part since I didn't know the reason it was there.

> Give me a few more days to play with this.

Let me know if you want me to make the changes. I think I understand the issue and can get it there.